gas-service/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+#![no_std]
+
+use crate::events::{
+    AddGasData, AddNativeGasData, GasPaidForContractCallData, NativeGasPaidForContractCallData,
+    RefundedData,
+};
+
+multiversx_sc::imports!();
+
+mod events;
+
+#[multiversx_sc::contract]
+pub trait GasService: events::Events {
+    #[init]
+    fn init(&self, gas_collector: &ManagedAddress) {
+        self.gas_collector().set_if_empty(gas_collector);
+    }
+
+    #[payable("*")]
+    #[endpoint(payGasForContractCall)]
+    fn pay_gas_for_contract_call(
+        &self,
+        destination_chain: ManagedBuffer,
+        destination_address: ManagedBuffer,
+        payload: ManagedBuffer,
+        refund_address: ManagedAddress,
+    ) {
+        let (gas_token, gas_fee_amount) = self.call_value().single_fungible_esdt();
+
+        let sender = self.blockchain().get_caller();
+        let hash = self.crypto().keccak256(&payload);
+
+        self.gas_paid_for_contract_call_event(
+            sender,
+            destination_chain,
+            destination_address,
+            GasPaidForContractCallData {
+                hash,
+                gas_token,
+                gas_fee_amount,
+                refund_address,
+            },
+        );
+    }
+
+    #[payable("EGLD")]
+    #[endpoint(payNativeGasForContractCall)]
+    fn pay_native_gas_for_contract_call(
+        &self,
+        destination_chain: ManagedBuffer,
+        destination_address: ManagedBuffer,
+        payload: ManagedBuffer,
+        refund_address: ManagedAddress,
+    ) {
+        let value = self.call_value().egld_value().clone_value();
+
+        require!(value > 0, "Nothing received");
+
+        let sender = self.blockchain().get_caller();
+        let hash = self.crypto().keccak256(&payload);
+
+        self.native_gas_paid_for_contract_call_event(
+            sender,
+            destination_chain,
+            destination_address,
+            NativeGasPaidForContractCallData {
+                hash,
+                value,
+                refund_address,
+            },
+        );
+    }
+
+    #[payable("*")]
+    #[endpoint(payGasForExpressCall)]
+    fn pay_gas_for_express_call(
+        &self,
+        destination_chain: ManagedBuffer,
+        destination_address: ManagedBuffer,
+        payload: ManagedBuffer,
+        refund_address: ManagedAddress,
+    ) {
+        let (gas_token, gas_fee_amount) = self.call_value().single_fungible_esdt();
+
+        let sender = self.blockchain().get_caller();
+        let hash = self.crypto().keccak256(&payload);
+
+        self.gas_paid_for_express_call(
+            sender,
+            destination_chain,
+            destination_address,
+            GasPaidForContractCallData {
+                hash,
+                gas_token,
+                gas_fee_amount,
+                refund_address,
+            },
+        );
+    }
+
+    #[payable("EGLD")]
+    #[endpoint(payNativeGasForExpressCall)]
```

Comment:
Why not make only one function and depending on the amount make it express or regular?